### ISSUE TITLE
Cached 403 access denied pages never refreshed if background-fetch is enabled

### ISSUE THREAD

**Issue by quicksketch** [backdrop/backdrop-issues#1805]
To reproduce:
- Enable the anonymous page cache (1 minute) and background fetch on the performance page.
- Create node/1. Mark as unpublished.
- As an anonymous user, visit node/1. You should get access denied.
- Edit node/1 as admin, publish the node.
- As the anonymous user, visit node/1. The first request you should still get a 403 access denied because of the background fetch (403 is the stale copy while the full version is being retrieved). The second request you should get the published node.
- On the second request (and all subsequent requests) you always get the 403 page.

No amount of waiting will ever show the published node for anonymous users. The page cache has to be manually cleared in order for the published node to show up.

**Comment by quicksketch** [comment]
The origin of this problem is in `_backdrop_bootstrap_page_cache()`:

```
      // Restore the metadata cached with the page.
      $_GET['q'] = $cache->data['path'];
      backdrop_set_title($cache->data['title'], PASS_THROUGH);
      date_default_timezone_set(backdrop_get_user_timezone());
```

On a path like `node/1` if the node is unpublished, Backdrop changes the internal path to `system/403` to display the 403 page. When the background fetch is enabled, the assignment of `$_GET['q'] = $cache->data['path'];` causes the current path to always be set to `system/403`. So the background-generated version of the page is always the 403 page instead of the path that is in the URL.

**Comment by quicksketch** [comment]
PR filed at https://github.com/backdrop/backdrop/pull/1327 with some new tests to check this doesn't happen again.

At first I thought we could just remove these lines entirely. They're there almost entirely for the sake of `hook_boot()` and `hook_exit()`, which are deprecated (and disabled by default) on cached page requests. However even after removing these lines, we had a different issue with the headers from `backdrop_serve_page_from_cache()` affecting the background-generated copy. So either way we needed to reset some static variables after serving the page.

I did move the setting of these variables to just before `hook_boot()` was called however, as there's no need to do things like set the timezone if we're just serving a cached page. This should give us a tiny performance improvement for anonymous page loads, as now we don't load `system.date.json` on cached pages.
